File: lib/src/main/kotlin/com/hdmain/tcpduplex/protocol/Protocol.kt
package com.hdmain.tcpduplex.protocol

import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.ByteOrder

object Protocol {
    const val MSG_TEXT: UByte = 1u
    const val MSG_PING: UByte = 2u
    const val MSG_PONG: UByte = 3u
    const val MSG_CLOSE: UByte = 4u

    private val HANDSHAKE_MAGIC = byteArrayOf('T'.code.toByte(), 'D'.code.toByte(), 'X'.code.toByte(), '1'.code.toByte())

    const val CURRENT_PROTOCOL_VERSION: UShort = 1u
    const val X25519_PUB_KEY_LEN = 32
    const val MAX_RECORD_PAYLOAD = 1 shl 20

    class BadHandshakeException(message: String = "protocol: invalid handshake") : Exception(message)
    class BadFrameException(message: String = "protocol: invalid frame") : Exception(message)
    class UnsupportedVersionException(val version: UShort) :
        Exception("protocol: unsupported protocol version: $version")

    fun supportsVersion(v: UShort): Boolean = when (v) {
        1.toUShort() -> true
        else -> false
    }

    private fun validateVersion(v: UShort) {
        if (!supportsVersion(v)) {
            throw UnsupportedVersionException(v)
        }
    }

    fun writeHandshake(out: OutputStream, negotiatedVersion: UShort, pubKey: ByteArray) {
        validateVersion(negotiatedVersion)
        out.write(HANDSHAKE_MAGIC)
        val verBuf = ByteBuffer.allocate(2).order(ByteOrder.BIG_ENDIAN)
        verBuf.putShort(negotiatedVersion.toShort())
        out.write(verBuf.array())
        if (pubKey.size != X25519_PUB_KEY_LEN) {
            throw BadHandshakeException()
        }
        out.write(pubKey)
    }

    fun readHandshake(input: InputStream): Pair<UShort, ByteArray> {
        val magic = readFull(input, HANDSHAKE_MAGIC.size)
        if (!magic.contentEquals(HANDSHAKE_MAGIC)) {
            throw BadHandshakeException()
        }
        val verBytes = readFull(input, 2)
        val protocolVersion = ByteBuffer.wrap(verBytes).order(ByteOrder.BIG_ENDIAN).short.toUShort()
        validateVersion(protocolVersion)
        val pubKey = readFull(input, X25519_PUB_KEY_LEN)
        return protocolVersion to pubKey
    }

    fun readRecord(input: InputStream): Pair<UByte, ByteArray> {
        val lenBuf = readFull(input, 4)
        val n = ByteBuffer.wrap(lenBuf).order(ByteOrder.BIG_ENDIAN).int.toUInt()
        if (n == 0u || n > MAX_RECORD_PAYLOAD.toUInt()) {
            throw BadFrameException()
        }
        val payload = readFull(input, n.toInt())
        val msgType = payload[0].toUByte()
        val sealed = payload.copyOfRange(1, payload.size)
        return msgType to sealed
    }

    fun writeRecord(out: OutputStream, msgType: UByte, sealed: ByteArray) {
        if (sealed.size > MAX_RECORD_PAYLOAD - 1) {
            throw BadFrameException()
        }
        val n = 1 + sealed.size
        val lenBuf = ByteBuffer.allocate(4).order(ByteOrder.BIG_ENDIAN)
        lenBuf.putInt(n)
        out.write(lenBuf.array())
        out.write(msgType.toInt())
        out.write(sealed)
    }

    private fun readFull(input: InputStream, size: Int): ByteArray {
        val buf = ByteArray(size)
        var off = 0
        while (off < size) {
            val n = input.read(buf, off, size - off)
            if (n < 0) {
                throw EOFException()
            }
            off += n
        }
        return buf
    }
}

File: tcpduplex/protocol/protocol.go
package protocol

import (
	"bytes"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
)

// Message types carried in the cleartext type byte (outside the ciphertext).
const (
	MsgText  uint8 = 1
	MsgPing  uint8 = 2
	MsgPong  uint8 = 3
	MsgClose uint8 = 4
)

const handshakeMagicLen = 4

var handshakeMagic = []byte{'T', 'D', 'X', '1'}

// CurrentProtocolVersion is the preferred protocol revision tcpduplex library implementations negotiate first.
const CurrentProtocolVersion uint16 = 1

// X25519PubKeyLen is the encoded length of an X25519 public key on the wire.
const X25519PubKeyLen = 32

// MaxRecordPayload is the maximum value of (1 + len(sealed)) on the wire.
const MaxRecordPayload = 1 << 20

var (
	ErrBadHandshake       = errors.New("protocol: invalid handshake")
	ErrBadFrame           = errors.New("protocol: invalid frame")
	ErrUnsupportedVersion = errors.New("protocol: unsupported protocol version")
)

// SupportsVersion reports whether v may be used as the negotiated wire revision for handshakes and framing.
func SupportsVersion(v uint16) bool {
	switch v {
	case 1:
		return true
	default:
		return false
	}
}

func validateVersion(v uint16) error {
	if SupportsVersion(v) {
		return nil
	}
	return fmt.Errorf("%w: %d", ErrUnsupportedVersion, v)
}

// WriteHandshake writes magic || negotiated protocol version || X25519 public key.
func WriteHandshake(w io.Writer, negotiatedVersion uint16, pubKey []byte) error {
	if err := validateVersion(negotiatedVersion); err != nil {
		return err
	}
	var hdr [2 + handshakeMagicLen]byte
	copy(hdr[:], handshakeMagic)
	binary.BigEndian.PutUint16(hdr[handshakeMagicLen:], negotiatedVersion)
	if _, err := w.Write(hdr[:]); err != nil {
		return err
	}
	if len(pubKey) != X25519PubKeyLen {
		return ErrBadHandshake
	}
	_, err := w.Write(pubKey)
	return err
}

// ReadHandshake reads and validates magic + declared peer protocol revision + public key.
func ReadHandshake(r io.Reader) (protocolVersion uint16, pubKey []byte, err error) {
	var magic [handshakeMagicLen]byte
	if err := readFull(r, magic[:]); err != nil {
		return 0, nil, err
	}
	if !bytes.Equal(magic[:], handshakeMagic) {
		return 0, nil, ErrBadHandshake
	}
	var ver [2]byte
	if err := readFull(r, ver[:]); err != nil {
		return 0, nil, err
	}
	protocolVersion = binary.BigEndian.Uint16(ver[:])
	if err := validateVersion(protocolVersion); err != nil {
		return 0, nil, err
	}
	pubKey = make([]byte, X25519PubKeyLen)
	if err := readFull(r, pubKey); err != nil {
		return 0, nil, err
	}
	return protocolVersion, pubKey, nil
}

func readFull(r io.Reader, buf []byte) error {
	_, err := io.ReadFull(r, buf)
	return err
}

// ReadRecord reads length || type || sealed (nonce||ciphertext||tag).
func ReadRecord(r io.Reader) (msgType uint8, sealed []byte, err error) {
	var lenBuf [4]byte
	if err := readFull(r, lenBuf[:]); err != nil {
		return 0, nil, err
	}
	n := binary.BigEndian.Uint32(lenBuf[:])
	if n == 0 || n > MaxRecordPayload {
		return 0, nil, ErrBadFrame
	}
	payload := make([]byte, n)
	if err := readFull(r, payload); err != nil {
		return 0, nil, err
	}
	msgType = payload[0]
	sealed = payload[1:]
	return msgType, sealed, nil
}

// WriteRecord writes length || type || sealed.
func WriteRecord(w io.Writer, msgType uint8, sealed []byte) error {
	if len(sealed) > MaxRecordPayload-1 {
		return ErrBadFrame
	}
	n := uint32(1 + len(sealed))
	var lenBuf [4]byte
	binary.BigEndian.PutUint32(lenBuf[:], n)
	if _, err := w.Write(lenBuf[:]); err != nil {
		return err
	}
	var hdr [1]byte
	hdr[0] = msgType
	if _, err := w.Write(hdr[:]); err != nil {
		return err
	}
	_, err := w.Write(sealed)
	return err
}
